flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java

```diff
@@ -0,0 +1,172 @@
+import java.util.concurrent.Executors;
+
+/**
+ * Data transfer utils for RocksDbKeyedStateBackend.
+ */
+class RocksDbStateDataTransfer {
+
+	public static void transferAllStateDataToDirectory(
+		IncrementalKeyedStateHandle restoreStateHandle,
+		Path dest,
+		int restoringThreadNum,
+		CloseableRegistry closeableRegistry) throws Exception {
+
+		final Map<StateHandleID, StreamStateHandle> sstFiles =
+			restoreStateHandle.getSharedState();
+		final Map<StateHandleID, StreamStateHandle> miscFiles =
+			restoreStateHandle.getPrivateState();
+
+		downloadDataForAllStateHandles(sstFiles, dest, restoringThreadNum, closeableRegistry);
+		downloadDataForAllStateHandles(miscFiles, dest, restoringThreadNum, closeableRegistry);
+	}
+
+	/**
+	 * Copies all the files from the given stream state handles to the given path, renaming the files w.r.t. their
+	 * {@link StateHandleID}.
+	 */
+	private static void downloadDataForAllStateHandles(
+		Map<StateHandleID, StreamStateHandle> stateHandleMap,
+		Path restoreInstancePath,
+		int restoringThreadNum,
+		CloseableRegistry closeableRegistry) throws Exception {
+
+		List<Runnable> runnables = createDownloadRunnables(stateHandleMap, restoreInstancePath, closeableRegistry);
+
+		if (isMultiThreadDownloadEnabled(restoringThreadNum)) {
+			runTransferWithMultipleThreads(runnables, restoringThreadNum, closeableRegistry);
+		} else {
+			runTransferWithSingleThread(runnables);
+		}
+	}
+
+	private static List<Runnable> createDownloadRunnables(
+		Map<StateHandleID, StreamStateHandle> stateHandleMap,
+		Path restoreInstancePath,
+		CloseableRegistry closeableRegistry) {
+		List<Runnable> runnables = new ArrayList<>(stateHandleMap.size());
+		for (Map.Entry<StateHandleID, StreamStateHandle> entry : stateHandleMap.entrySet()) {
+			StateHandleID stateHandleID = entry.getKey();
+			StreamStateHandle remoteFileHandle = entry.getValue();
+
+			Path path = new Path(restoreInstancePath, stateHandleID.toString());
+
+			runnables.add(ThrowingRunnable.unchecked(new ThrowingRunnable<IOException>() {
+				@Override
+				public void run() throws IOException {
+					downloadDataForStateHandle(path, remoteFileHandle, closeableRegistry);
+				}
+			}));
+		}
+		return runnables;
+	}
+
+	private static void runTransferWithMultipleThreads(
+		List<Runnable> runnables,
+		int threadNum,
+		CloseableRegistry closeableRegistry) throws ExecutionException, InterruptedException, IOException {
+
+		final ExecutorService executorService = Executors.newFixedThreadPool(threadNum);
+		Closeable shutDownAllTasksAndExecutorService = () -> executorService.shutdownNow();
+		closeableRegistry.registerCloseable(shutDownAllTasksAndExecutorService);
+
+		try {
+			List<CompletableFuture<Void>> futures = new ArrayList<>(runnables.size());
+			for (Runnable runnable : runnables) {
+				CompletableFuture<Void> future = CompletableFuture.runAsync(runnable, executorService);
+				futures.add(future);
+			}
+			FutureUtils.waitForAll(futures).get();
```

Comment:
Let's remove the `ExecutionException` by unwrapping it via `ExceptionUtils.stripExecutionException`